### ISSUE TITLE
[feature request] RNN hidden weight for Multi GPU training

### ISSUE THREAD

**Issue by wangkenpu** [pytorch/pytorch#7890]
As we know, the input and output tensor support ```batch_first``` for RNN training, but as for hidden state, the tensor shape is forced to be ```(num_layers * num_directions, batch, hidden_size)```, even we set ```batch_first = True```. So there is a problem when we do multi GPU training by using ```torch.nn.DataParallel```. 

For example, if we initialize a hidden state with shape ```(num_layers * num_directions, batch, hidden_size)``` and set ```hidden = init_hidden(batch_size)``` like [word_language_model](https://github.com/pytorch/examples/blob/f9820471d615d848c14661b2d582417ca3aee8a3/word_language_model/main.py#L150)，then if we set ```hidden = init_hidden(batch_size)``` and use parallel training, the hidden size will be ```(num_layers * num_directions / num_gpu, batch, hidden_size)```  while the correct shape is ```(num_layers * num_directions, batch / num_gpu, hidden_size)```. So I was wondering if PyTorch Teams prepare to support ```batch_first``` for RNN hidden state.

Another possible way to avoid this problem is to use the following codes:
```
def init_hidden(self, batch_size):
    weight = next(self.parameters())
    # The hidden weight format is not consisten with PyTorch's LSTM
    # impelementation, so we will transpose it
    hidden = weight.new_zeros(batch_size, self.num_layers * 2,
                              self.hidden_size, requires_grad=False)
    hidden = [hidden, hidden]
    return hidden

def forward(self, inputs, hidden, length):
    hidden[0] = hidden[0].permute(1, 0, 2).contiguous()
    hidden[1] = hidden[1].permute(1, 0, 2).contiguous()
    inputs_pack = pack_padded_sequence(inputs, length, batch_first=True)
    # self.blstm.flatten_parameters()
    output, hidden = self.blstm(inputs_pack, hidden)
    ......
    hidden = list(hidden)
    hidden[0] = hidden[0].permute(1, 0, 2).contiguous()
    hidden[1] = hidden[1].permute(1, 0, 2).contiguous()
    return output, hidden
```

But this implementation is very ugly. Also, this implementation faces another problem like the following warning log:
```
/mnt/workspace/pytorch/deep_clustering/model/blstm_upit.py:65: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters().
```
If we add ```self.blstm.flatten_parameters()``` before ```output, hidden = self.blstm(inputs_pack, hidden)```, we will face another problem like #7092

**Comment by SsnL** [comment]
Set `dim=1` in `nn.DataParallel` https://pytorch.org/docs/master/nn.html#torch.nn.DataParallel .

**Comment by wangkenpu** [comment]
Thanks. However, when do RNN training and set ```batch_first = True```, setting ```dim = 1``` in ```nn.DataParallel``` is not useful. 

Input of shape: ```(batch_size, seq_len, input_size)```
Hidden of shape: ```(num_layers * num_directions, batch, hidden_size)```

When set ```dim = 1``` in ```nn.DataParallel```, it only can chunk the hidden state correctly, while input is chunked in the ```seq_len``` dimension.

**Comment by SsnL** [comment]
@wangkenpu I see. You have a valid point. I'm reopening this to see if anyone wants to discuss per-input `dim` arg for `DataParallel`.

I am wondering if you are just doing one recurrent cell in each forward? If that's the case, do you really need multi-GPU? If each forward is an entire sequence, how about initializing `hidden` in forward (or use default `hidden` that is init'ed to zero if you are using the provided RNN layers)?

**Comment by ngimel** [comment]
While the point about DataParallel and hidden splitting inconsistencies is valid, for RNNs especially you are usually much better off using DistributedDataParallel with 1 GPU per process, and DistributedDataParallel does not suffer from this as it does not need to split input.

**Comment by wangkenpu** [comment]
@SsnL In my experiments, I often use LSTM with more than 300 recurrent cells, and if I use BPTT, the hidden states are very important, while default ```hidden``` is only valid for entire sequence training. 

In my opinion, maybe we don't need to implement per-input  ```dim``` arg for ```DataParallel```.  We just need to make ```hidden``` in RNN support ```batch_first```.

**Comment by lijianhackthon** [comment]
Have faced exactly the same issue with @wangkenpu . Hope the Pytorch team can add this feature. Stay tuned.

**Comment by Hshihua** [comment]
@wangkenpu The same issue with you. Have you found a valid solution to address such a problem?

**Comment by wangkenpu** [comment]
1. try to use nn.LSTMCell instead of nn.LSTM
2. try to redefine hidden shape and transpose it in each forward step
